### DISCUSSION TITLE
[New Feature][1.1.216] Integrated openpose editor

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#1548] [General]
Sometimes the controlnet openpose preprocessor does not provide exact result we want. When this happens, we can either:
- Do a pose edit from 3rd party editors such as posex, and use that as input image with preprocessor `none`
  - Cons: Existing extensions have bad/no support for hand/face. They might not receive the most up to date pose detection code from ControlNet, as most of them copy a version of ControlNet's pose detection code.
- Download the json file provided by controlnet preview and try to correct it in a 3rd party editor
  - Cons: many steps of operation needed: download a file, upload to another tool, download the updated pose back, set it in the input image, choose preprocessor `none`.

To address this issue. We implemented an embedded openpose editor. Now if you are not satisfied with the pose output, you can click the `Edit` button on the generated image to send the pose to an editor for edit.  

Currently, to use the edit feature, you will need controlnet `v1.1.216` and another extension installed:
https://github.com/huchenlei/sd-webui-openpose-editor. 
(Note: although the extension is build with Vue.js, the users do not need to setup the build environment. The extension will download the compiled app `dist/` from github release.)

Here is a demo showing how to use the feature:
[edit_demo.webm](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/4fa1addf-a30a-42df-8518-8a71d3d276a5)

If you don't like the feature, here is a guide on how to disable it:
[feature_off_demo.webm](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/b7281c27-7445-41ef-b0c2-33f9e4a184a4)

If you meet any problems with the editor, please submit an issue to the editor repo instead of the controlnet extension repo, thanks!

**Comment by chekaaa** [comment]
Should I install the openpose editor using the extension tab ? because I did and still show the message that I dont have the openpose editor installed.

**Reply by huchenlei** [reply]
You need to restart the WebUI commandline after installing the extension as it will mount the editor to `openpose_editor_index` on A1111 start. Clicking restart in the WebUI might not be sufficient.

**Reply by yuange250** [reply]
这个有用

**Comment by Woisek** [comment]
> Here is a demo showing how to use the feature: [edit_demo.webm](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282

Could you please provide a video in a broadly available format like MP4? Thanks.

**Reply by anonghuser** [reply]
Rename the WebM to MKV and you have a broadly available format!

**Comment by Woisek** [comment]
Seems not quite there yet ... 🤔
![Zwischenablagebild](https://github.com/Mikubill/sd-webui-controlnet/assets/122026301/792542d2-a8de-498b-a6fc-cb22afdc28fa)

**Reply by huchenlei** [reply]
Sorry mybad the modal background is not adjusting based on the theme but texts are. 

There is a line of prompt showing that the required extension is not installed.

**Reply by chekaaa** [reply]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/41851841/b55b1a4a-4136-4441-a1b6-39dc051fcfaa)
I got the same message , and I do have it installed and restarted too. I feel like I'm missing a step

**Reply by Woisek** [reply]
Are there more "openpose editors"? I have this one installed -> https://github.com/fkunn1326/openpose-editor

**Reply by chekaaa** [reply]
I do have more but disabled , the only one enables is the one recommended by controlnet https://github.com/huchenlei/sd-webui-openpose-editor.

**Reply by huchenlei** [reply]
Do you get following trace in the commandline output?

```
Traceback (most recent call last):
  File "D:\fresh_chekouts\stable-diffusion-webui\modules\script_callbacks.py", line 108, in app_started_callback
    c.callback(demo, app)
  File "D:\fresh_chekouts\stable-diffusion-webui\extensions\sd-webui-openpose-editor\scripts\openpose_editor.py", line 99, in mount_openpose_api
    StaticFiles(directory=DIST_DIR, html=True),
  File "D:\fresh_chekouts\stable-diffusion-webui\venv\lib\site-packages\starlette\staticfiles.py", line 57, in __init__
    raise RuntimeError(f"Directory '{directory}' does not exist")
RuntimeError: Directory 'extensions/sd-webui-openpose-editor/dist' does not exist
```

**Reply by huchenlei** [reply]
Seems like the auto download release code is not working properly. But you can download `dist.zip` here:
https://github.com/huchenlei/sd-webui-openpose-editor/releases/tag/v0.0.7 and unzip it under the editor extension folder. I am working on fixing the autoupdate script. Sorry about that.

**Reply by chekaaa** [reply]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/41851841/a0c7322f-4b71-42c7-87c4-b8ba5e0026a4)
I doesnt show that  trace

**Reply by huchenlei** [reply]
Fix: [9768dab](https://github.com/huchenlei/sd-webui-openpose-editor/commit/9768dab416d7e87820f5527aed54432e97a4422c)

Was using wrong unzip dir. The autoupdate should work now.

**Reply by Woisek** [reply]
OK, now it's working. 👍
But now, at first sight, I have noticed the scourge that affects any extension that works with "details": it lacks a zoom function. 😁
Zoom is indispensable if we want to move the points exactly. I hope you already have something in work ... 😄

**Reply by huchenlei** [reply]
Well, the full zoom is not supported right now but on the devlist. Currently we do have an autozoom feature. If you expand face/hand component, the canvas will auto zoom to the object.

**Reply by huchenlei** [reply]
> OK, now it's working. 👍 But now, at first sight, I have noticed the scourge that affects any extension that works with "details": it lacks a zoom function. 😁 Zoom is indispensable if we want to move the points exactly. I hope you already have something in work ... 😄

@Woisek 
Full zoom feature added: https://github.com/huchenlei/sd-webui-openpose-editor/releases/tag/v0.0.8

**Reply by huchenlei** [reply]
Panning feature added: https://github.com/huchenlei/sd-webui-openpose-editor/releases/tag/v0.0.10

**Reply by Woisek** [reply]
Wow, this was fast! 😳
Maybe I'm blind, but, how can we zoom/pan ... ? Are there shortcut for this?
Thanks for this great addition! 👍

**Reply by huchenlei** [reply]
Zoom using mouse scroll and pan by dragging on canvas with SPACE key down.

**Reply by Woisek** [reply]
Thanks for the reply, but ... scrolling scrolls the page and as soon I press SPACE, the page scrolls again down. 😕
I'm not sure if the space bar is a good choice. Are you aware of canvas zoom? -> https://github.com/richrobber2/canvas-zoom.git
Maybe you can adapt his shortcuts for the pan and zoom. Maybe even do it exactly, so we could use the same shortcuts on both extensions and wouldn't have to rethink every time. 😊

**Reply by huchenlei** [reply]
Fixed in v0.0.11. The default scrolling behaviour of pressing space is prevented. I know people might have different preference on which key to press. Maybe I will add a setting for keybindings if the list keeps growing. For now I would prefer SPACE as it matches the keybinding of Photoshop.

**Reply by Woisek** [reply]
Sorry to say but pressing space still triggers scrolling ... 😔

**Reply by huchenlei** [reply]
I cannot reproduce it locally. Can you submit an issue along with a video showing the behaviour? https://github.com/huchenlei/sd-webui-openpose-editor/issues

**Reply by Woisek** [reply]
I can submit the issue but a video ... ? When pressing the space bar, the page scrolls down. Nothing more complex about it. 😊
And unfortunately I have nothing to record this, so ... I can just tell you what's going on.

**Comment by lybmccree** [comment]
640 * 960的图放入controlnet,勾选pixel perfect，点击预览, 然后Edit，打开 窗口，他的512 * 768分辨率，然后我再上传这张素材当背景，完全对不上，可以让素材图片自动 缩到512*768吗

**Reply by lybmccree** [reply]
现在只能去掉勾选pixel perfect，然后手动设置预处理器的分辨率为640，然后生成的图才是640

**Reply by huchenlei** [reply]
背景图如果自己上传的话现在就只能通过改scale ratio来改变大小。fabricjs很奇怪图不让用control point操控。
现在只有背景图是从ControlNet直接传过来的时候会自动进行缩放。
你要是有截图的话能发一下吗？能让我更好了解下问题在哪

**Reply by lybmccree** [reply]
复现步骤，
1. 浏览器新打开一个窗口页面，保证各种ControlNet参数都是默认
2. 用一张640x960的图放入ControlNet
3. 勾选PixelPerect，勾选Allow Preview
4. 选OpenPose预处理器，然后 点击 预览 按钮
5. 然后点击 预览图上面的Edit
6. 在OpenPose Editor界面点击 Upload Image按钮就能复现了

**Reply by huchenlei** [reply]
好像有点不对，背景图是你自己上传的吗？如果是ControlNet点edit直接传过来的图，图的名字会是`controlnet input`，而且默认是lock状态。

你这边是点击edit没有从ControlNet那边自动带背景图过来吗？

**Reply by lybmccree** [reply]
我更新到最新，从来没试过有自动带背景过来的呀。难道配置不对，我就拉了代码，装了一下nodejs，没配置啥环境，然后重启SD就有这个功能了，用的秋叶5.2稳定版本

**Reply by lybmccree** [reply]
用的https://github.com/huchenlei/sd-webui-openpose-editor 这个代码

**Reply by huchenlei** [reply]
https://github.com/Mikubill/sd-webui-controlnet/blob/f36493878b299c367bc51f2935fd7e6d19188569/javascript/openpose_editor.js#L34

你试试点开Chrome Devtools看一下有没有报错；没报错的话看看打断点在这一行看看URL有没有拿到值。正常这会是一个base64的data url。

**Reply by huchenlei** [reply]
打断点要点到Source，然后左边文件目录localhost > file=extensions > sd-webui-controlnet > openpose_editor.js

**Reply by huchenlei** [reply]
另外其实不用装node，插件是直接去GitHub下载dist不需要自己compile的。

**Reply by lybmccree** [reply]
不装node，直接就打不开，有问题，打开显示缺少 indedx什么鬼的,重启3次SD都没用，我是直接git clone的代码

**Reply by huchenlei** [reply]
嗯？有点奇怪啊，这个缺少index是前端Chrome里报的错还是commandline后端报的错？

**Reply by lybmccree** [reply]
网页里面点了预览后，点击Edit按钮 打开那个OpenPose Editor窗口就显示缺少东西，出不来东西，有可能我几次都下载失败那个dist的吧

**Reply by lybmccree** [reply]
用的Edge，控制台报错
index-5f592b6b.js:583  Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'digest')
    at ape (index-5f592b6b.js:583:27591)
    at Proxy.loadCanvasFromFrameMessage (index-5f592b6b.js:583:35841)
    at index-5f592b6b.js:583:29928
ape @ index-5f592b6b.js:583
loadCanvasFromFrameMessage @ index-5f592b6b.js:583
(匿名) @ index-5f592b6b.js:583
await in (匿名)(异步)
(匿名) @ index-5f592b6b.js:583
postMessage(异步)
(匿名) @ openpose_editor.js?1685930843.5780058:32

**Reply by huchenlei** [reply]
https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest

这个edge应该是支持的啊，2020年就支持了。你换Chrome试试？

**Reply by lybmccree** [reply]
谷歌浏览器也不行,一样的报错

**Reply by lybmccree** [reply]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/42838946/e7391d98-1624-413e-8ceb-8afb5ad3b82b)
新安装的谷歌浏览器，电脑也是上个月买的

**Reply by huchenlei** [reply]
我能想到的问题就是有其他code把crypto.subtle给overwrite掉了。你直接在控制台里打crypto.subtle看看是不是undefined

**Reply by lybmccree** [reply]
是的undefined,这就奇怪了

**Reply by huchenlei** [reply]
新开一个不是sd的页测一下是不是？正常web api在空页面上是肯定存在的。应该是其他插件把全局crypto这个变量名给overwrite掉了

**Reply by lybmccree** [reply]
我在网上搜了一下，不使用127.0.0.1启动的，crypto.subtle就有问题，因为这边是做成局域网了，用了192开头，之前用nginx代理，秋叶更新后就不能用nginx代理了

**Reply by huchenlei** [reply]
https://stackoverflow.com/questions/46468104/how-to-use-subtlecrypto-in-chrome-window-crypto-subtle-is-undefined ok那我看看能不能不调crypto算hash

**Reply by lybmccree** [reply]
好的，辛苦了

**Reply by huchenlei** [reply]
Fixed. Now we are using a 3rd party library to do hashing intead of relying on Web Crypto API.

**Reply by sdcarterchen** [reply]
> 好像有点不对，背景图是你自己上传的吗？如果是ControlNet点edit直接传过来的图，图的名字会是`controlnet input`，而且默认是lock状态。
> 
> 你这边是点击edit没有从ControlNet那边自动带背景图过来吗？

我试着，大部分情况下pose和背景图对不上，需要调那个scale ratio去对上。还有个问题就是，手部分的手指颜色左右手表示的是同一个手指头吗？我看open pose自动生成的和这个插件里面显示的调整节点对不上

**Reply by huchenlei** [reply]
可以交一个issue到https://github.com/huchenlei/sd-webui-openpose-editor 吗？最好附上截图和复现的图片/json文件。

**Comment by katopz** [comment]
Good job! Looking forward to this.👍
BTW I'm curious why you chose [Openpose over MediaPipe](https://www.hearai.pl/post/14-openpose/)?

**Reply by huchenlei** [reply]
From the editor perspective, media pipe's face representation has too many keypoints, which is not suitable for manual edits. Openpose at basic body level provides 5 keypoints for head (nose, L/R eye, L/R ear), which is easy to manage.

**Comment by 2blackbar** [comment]
Very good for editing already clean armatures but It needs a way to spawn missing hand instead of clicking points one by one, also premade face points for side,front,4/3 view, thats most used ones, like a library of premade json files ,theres not that many of them really for the most often used angles.
I mean yeah i guess i can make a library and save the hand poses myself but a button "spawn hand" or "spawn face" would be handy, these are the ones we will be fixing  most often
Red highlighting on the list with names for the bone im grabbing would be handy , that shouldnt be too hard cause i can see the numbers changing when i move the point, so id make the number red and also its name red when its being moved
what up with black font on black background ?
Clone and flip hand?

Sometimes i get lost /duplicate dots of hands leftovers and cant get rid of them, oh it also happened with face, yeah i think this needs some checks, or a delete button
Wit the missing fingers, i think it wouldnt be THAT hard to approximate position of the missing fingers when theres already at least one finger ? it doesnt have to precise, now i have to manyually input where finger should be and position is kinda easy to figure out with code imo if you have at least one saved hand json file, then we can clean it up after all missing fingers will appear.Id rather have hand restored in wrong position than spend time adding bones one by one 
Theres a bug where it will recognize both faces on image with 2 people, once i go into edit, the second face disappears
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/33694664/69de36f2-3cf0-422f-8999-f2cc8cec976c)

**Reply by huchenlei** [reply]
Really appreciate your feedback!

There is already a spawn default hand feature. If you click "x" to remove a hand you don't want, a "Add Left/Right Hand" button will appear. Click it will add the default hand skeleton to the canvas.

Spawn face is not supported as I think 70 keypoints on face makes it very hard to adjust manually, but you can always upload a JSON file, and the editor will load the face in the JSON file to the person you selected. The same also applies for hands. 

The highlighting of active keypoints is achievable if you find it useful. You can submit a feature request issue in the editor repo.

Currently there is a "fire" symbol if there are invalid keypoints detected for the object. Clicking it will unjam all keypoints, i.e. align them as a line on the canvas so that you can move them to desired location one by one. 

For the bug, can you submit a bug report in the editor repo? If you can attach the problematic JSON in the report that would be fantastic. 

Thanks!

**Reply by 2blackbar** [reply]
oh thanks, but how to mirror left hand right hand? i think ab utton for that would be handy, i got the hand spawned but i have to mirror it 
Its pretty hard to scale the hand once its spawned, i think corners arent reacting that well to drag
But yeah this is already very good

**Reply by huchenlei** [reply]
There is a section in README explaining group vs selection, and a video demo on the difference. In fabricjs api, grouped objects cannot be scaled or rotated. So what you need to do is ungroup the hand, select the hand keypoints by drag on canvas, and you should be able to scale/rotate/mirror the selection. 

I know this is awkward and counter intuitive. I am also working on a solution to it, but for now under my current setup, the grouped object is only moveable.

**Reply by 2blackbar** [reply]
yeah it has its workflow i need to adapt to, i flip the hand once im in hand edit mode cause when i flip it before that then it explodes when going to edit mode.
Imo the editor is quite complete, i was able to fix the hand adding it and moving fingers pretty easily, You did a very good job
Can you lock selecting and moving non hand points when we are in hand editing mode? i dont really want to move wrist or anythin else but hand

**Reply by huchenlei** [reply]
Grouping is the mechanism for lock movement for other keypoints. You click the icon right besides the eye icon, and the object will be grouped. Once grouped, you cannot operate on individual keypoints within the group.

**Reply by huchenlei** [reply]
v0.1.2 implemented many features you requested:
- Selected keypoints will be highlighted in the control panel
- Body parts will be lockable
- Grouped objects can be scaled and rotated

**Comment by zxczjg123** [comment]
GitCommandError: Cmd('git') failed due to: exit code(128) cmdline: git clone -v --filter=blob:none -- https://github.com/Mikubill/sd-webui-controlnet/discussions/1548 D:\sd-webui-aki-v4.3\tmp\1548 stderr: 'Cloning into 'D:\sd-webui-aki-v4.3\tmp\1548'... warning: missing OAuth configuration for gitcode.net - see https://aka.ms/gcm/gitlab for more information remote: HTTP Basic: Access denied fatal: Authentication failed for 'https://gitcode.net/ranting8323/sd-webui-controlnet/discussions/1548.git/' '
